File: tests/simple_ops/decrypt.cpp
// Copyright 2024-present Niobium Microsystems, Inc.
// Licensed under the Apache License, Version 2.0.
//
// Simple ops example — decrypt and verify (CKKS)
//
// Usage: ./simple_ops_decrypt [key_dir [operation]]

#include "openfhe.h"

#include <cmath>

#include "ciphertext-ser.h"
#include "cryptocontext-ser.h"
#include "key/key-ser.h"
#include "scheme/ckksrns/ckksrns-ser.h"

using namespace lbcrypto;

int main(int argc, char* argv[]) {
    std::string keyDir = "simple_ops_keys";
    std::string operation = "ADD";
    std::string ct_file = "ct_result.bin";  // override with 3rd arg

    if (argc > 1) keyDir = argv[1];
    if (argc > 2) operation = argv[2];
    if (argc > 3) ct_file = argv[3];

    std::cout << "=== Simple Ops — Decrypt (" << operation
              << ", ct=" << ct_file << ") ===" << std::endl;

    CryptoContext<DCRTPoly> cc;
    if (!Serial::DeserializeFromFile(keyDir + "/cc.bin", cc, SerType::BINARY))
        throw std::runtime_error("Failed to load crypto context");

    PrivateKey<DCRTPoly> secretKey;
    if (!Serial::DeserializeFromFile(keyDir + "/sk.bin", secretKey, SerType::BINARY))
        throw std::runtime_error("Failed to load secret key");

    Ciphertext<DCRTPoly> ct_result;
    if (!Serial::DeserializeFromFile(keyDir + "/" + ct_file, ct_result, SerType::BINARY))
        throw std::runtime_error("Failed to load result ciphertext: " + ct_file);

    double a = 0, b = 0;
    {
        std::ifstream valStream(keyDir + "/values.txt");
        valStream >> a >> b;
    }

    Plaintext pt_result;
    cc->Decrypt(secretKey, ct_result, &pt_result);
    pt_result->SetLength(1);

    double result = pt_result->GetRealPackedValue()[0];

    // Compute expected value based on operation
    double expected = 0;
    if (operation == "ADD")          expected = a + b;
    else if (operation == "SUB")     expected = a - b;
    else if (operation == "MUL")     expected = a * b;
    else if (operation == "NEG")     expected = -a;
    else if (operation == "ADDI")    expected = a + 3.0;
    else if (operation == "SUBI")    expected = a - 2.0;
    else if (operation == "MULI")    expected = a * 4.0;
    else if (operation == "ADD_ADD") expected = 2 * a + b;
    else if (operation == "ADD_SUB") expected = b;
    else if (operation == "MUL_ADD") expected = a * b + a;
    else if (operation == "ADD_MUL") expected = (a + b) * a;
    else if (operation == "MUL_MUL") expected = a * b * a;
    else if (operation == "ALL_NO_MUL") expected = (b + 1.0) * 4.0;
    else if (operation == "MORPH")   expected = b;  // EvalRotate(ct_a={a,b}, 1)[0]
    else {
        std::cerr << "Unknown operation: " << operation << std::endl;
        return 1;
    }

    double diff = std::abs(result - expected);
    double tolerance = 0.01;

    std::cout << "Result: " << result << " (expected " << expected
              << ", diff " << diff << ")" << std::endl;

    if (diff < tolerance) {
        std::cout << "[PASS] " << operation << ": " << result << " ~= " << expected << std::endl;
        return 0;
    } else {
        std::cerr << "[FAIL] " << operation << ": " << result << " != " << expected
                  << " (diff " << diff << " > " << tolerance << ")" << std::endl;
        return 1;
    }
}

File: python/tests/simple_ops/decrypt.py
#!/usr/bin/env python3
"""simple_ops decrypt — Python port of tests/simple_ops/decrypt.cpp.

Decrypts the result ciphertext and checks slot 0 against the expected value for
the op. Usage: decrypt.py <dir> <op> [ct_file].
"""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from _nbcommon import o, BIN  # noqa: E402

# Expected slot-0 value per op, matching decrypt.cpp.
EXPECTED = {
    "ADD": lambda a, b: a + b,
    "SUB": lambda a, b: a - b,
    "MUL": lambda a, b: a * b,
    "NEG": lambda a, b: -a,
    "ADDI": lambda a, b: a + 3.0,
    "SUBI": lambda a, b: a - 2.0,
    "MULI": lambda a, b: a * 4.0,
    "ADD_ADD": lambda a, b: 2 * a + b,
    "ADD_SUB": lambda a, b: b,
    "MUL_ADD": lambda a, b: a * b + a,
    "ADD_MUL": lambda a, b: (a + b) * a,
    "MUL_MUL": lambda a, b: a * b * a,
    "ALL_NO_MUL": lambda a, b: (b + 1.0) * 4.0,
    "MORPH": lambda a, b: b,
}


def main(argv):
    d = argv[1]
    op = argv[2]
    ct_file = argv[3] if len(argv) > 3 else "ct_result.bin"

    cc, ok = o.DeserializeCryptoContext(f"{d}/cc.bin", BIN)
    assert ok, "cc"
    sk, ok = o.DeserializePrivateKey(f"{d}/sk.bin", BIN)
    assert ok, "sk"
    ct, ok = o.DeserializeCiphertext(f"{d}/{ct_file}", BIN)
    assert ok, ct_file
    with open(f"{d}/values.txt") as fh:
        a, b = (float(x) for x in fh.read().split())

    pt = cc.Decrypt(ct, sk)
    pt.SetLength(1)
    got = pt.GetRealPackedValue()[0]
    exp = EXPECTED[op](a, b)
    ok = abs(got - exp) < 0.01
    print(f"[{'PASS' if ok else 'FAIL'}] {op}: {got:.4f} ~= {exp:.4f}")
    sys.exit(0 if ok else 1)


if __name__ == "__main__":
    main(sys.argv)
